calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,489 @@
+            });
+            cur_state
+        }
+        ir::Control::If(ir::If {
+            tbranch,
+            fbranch,
+            attributes,
+            ..
+        }) => {
+            attributes.insert(BEGIN_ID, cur_state);
+            cur_state += 1;
+            cur_state = compute_unique_ids(tbranch, cur_state);
+            cur_state = compute_unique_ids(fbranch, cur_state);
+            attributes.insert(END_ID, cur_state);
+            cur_state + 1
+        }
+        ir::Control::While(ir::While {
+            body, attributes, ..
+        }) => {
+            attributes.insert(NODE_ID, cur_state);
+            cur_state += 1;
+            cur_state = compute_unique_ids(body, cur_state);
+            cur_state
+        }
+        ir::Control::Empty(_) => cur_state,
+    }
+}
+
+impl DominatorMap {
+    /// Construct a domination map.
+    pub fn new(control: &mut ir::Control, component_name: String) -> Self {
+        compute_unique_ids(control, 0);
+        let mut exits_map = HashMap::new();
+        Self::build_exit_map(control, &mut exits_map);
+        let mut map = DominatorMap {
+            map: HashMap::new(),
+            exits_map,
+            component_name,
+        };
+        Self::build_map(control, &mut map);
+        map
+    }
+
+    //Builds the domination map by running update_map() until the map
+    //stops changing.
+    fn build_map(main_c: &ir::Control, d_map: &mut DominatorMap) {
+        let mut og_map = d_map.map.clone();
+        let empty_set: HashSet<u64> = HashSet::new();
+        Self::update_map(main_c, 0, &empty_set, d_map);
+        while og_map != d_map.map {
+            og_map = d_map.map.clone();
+            Self::update_map(main_c, 0, &empty_set, d_map);
+        }
+    }
+
+    //Given a control, gets its associated id. For if statments, gets the
+    //beginning id if begin_id is true and end_id if begin_id is false.
+    //Should not be called on empty control
+    //statements or any other statements that don't have an id numbering.
+    fn get_id(c: &ir::Control, begin_id: bool) -> u64 {
+        if let Some(v) = match c {
+            ir::Control::If(_) => {
+                if begin_id {
+                    get_attr(c, BEGIN_ID)
+                } else {
+                    get_attr(c, END_ID)
+                }
+            }
+            _ => get_attr(c, NODE_ID),
+        } {
+            v
+        } else {
+            unreachable!(
+                "get_id() shouldn't be called on control stmts that don't have id numbering"
+            )
+        }
+    }
+
+    //given a control stmt c and a key, returns true if c matches key, false
+    //otherwise. For if stmts return true if key matches either begin or end id.
+    fn matches_key(c: &ir::Control, key: u64) -> bool {
+        let mut ids = vec![Self::get_id(c, true)];
+        if let Some(end) = get_attr(c, END_ID) {
+            ids.push(end);
+        }
+        ids.contains(&key)
```

Comment:
Why construct a vector and call contains on it? It seems we'll only ever have `key == end` or `key == Self::get_id(c, true)`. Just do those comparisons directly.